cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -202,24 +203,102 @@ protected void addPartition(PartitionKey key, String path, boolean explorable, M
 
   @Override
   public PartitionConsumerResult consumePartitions(PartitionConsumerState partitionConsumerState) {
+    return consumePartitions(partitionConsumerState, PartitionFilter.ALWAYS_MATCH, Integer.MAX_VALUE);
+  }
+
+  /**
+   * While applying a partition filter and a limit, parse partitions from the rows of a scanner and add them to a list.
+   * Note that multiple partitions can have the same transaction write pointer. For each set of partitions with the same
+   * write pointer, we either add the entire set or exclude the entire set. The limit is applied after adding each such
+   * set of partitions to the list.
```

Comment:
Yea, I described the reason and alternative here:
https://issues.cask.co/browse/CDAP-3102